File: Problems/Algorithms/77. Combinations/Combinations.kt
class Solution {
    private val curr: MutableList<Int>
    private val res: MutableList<MutableList<Int>>
    
    init {
        curr = mutableListOf()
        res = mutableListOf()
    }
    
    fun combine(n: Int, k: Int): List<List<Int>> {
        aux(n, k, 1)
        return res
    }
    
    private fun aux(n: Int, k: Int, start: Int): Unit {
        if (curr.size == k) {
            res.add(ArrayList(curr))
            return
        }
        
        for (i in start..n) {
            curr.add(i)
            aux(n, k, i+1)
            curr.removeAt(curr.size - 1)
        }
    }
}

File: Problems/Algorithms/77. Combinations/combinations.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> ans;
    vector<int> curr;

    void helper(int i, int n, int k, vector<int>& curr) {
        if (k == 0) {
            ans.push_back(curr);
            return;
        }

        if (i == n) {
            return;
        }

        helper(i+1, n, k, curr);
        curr.push_back(i+1);
        helper(i+1, n, k-1, curr);
        curr.pop_back();
    }

    vector<vector<int>> combine(int n, int k) {
        helper(0, n, k, curr);
        return ans;
    }
};
